tree/30_generic_methods/user_secondary_group.cf

```diff
@@ -0,0 +1,160 @@
+      "subcall_prefix[${group_list}]" string => canonify("user_group_${login}_${group_list}");
+      "prefixes"                       slist => { "${class_prefix}",
+                                                  "${report_data.method_id}"
+                                                };
+      "c_id"                          string => canonify("${report_data.method_id}");
+
+
+  defaults:
+      "force" string => "false", if_match_regex => "";
+
+  classes:
+      "user_exists"         expression => userexists("${login}");
+      "group_not_found"            not => groupexists("${group_list}");
+      "purge_other_groups"  expression => strcmp("${force}", "true");
+
+      "should_report"       expression => "${report_data.should_report}";
+
+    any::
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+
+    pass3::
+      "error_subcall"    expression => "${subcall_prefix[${group_list}]}_error";
+      "repaired_subcall" expression => "${subcall_prefix[${group_list}]}_repaired";
+      "success_subcall"  expression => and( "${subcall_prefix[${group_list}]}_kept",
+                                            not("error_subcall|repaired_subcall")
+                                          );
+
+      # ${c_id}_{error|repaired|success} are already defined if in enforce
+      # There are redefined here for clarity
+      "${c_id}_error"    expression => or( "!user_exists",
+                                           "group_not_found",
+                                           "error_subcall",
+                                           "${c_id}_error"
+                                         );
+
+      "${c_id}_repaired" expression => and( "user_exists",
+                                            "!group_not_found",
+                                            or( "repaired_subcall",
+                                                "${c_id}_repaired"
+                                              )
+                                          );
+
+      "${c_id}_success"  expression => and( "user_exists",
+                                            "!group_not_found",
+                                            or( "success_subcall",
+                                                "${c_id}_success"
+                                              )
+                                          );
+
+  users:
+    pass2::
+      "${login}"
+        policy => "present",
+        groups_secondary => { @{group_list} },
+        classes => classes_generic_two("${class_prefix}", "${report_data.method_id}"),
+        if => and( "user_exists",
+                   "!group_not_found",
+                   "purge_other_groups"
+                 );
+
+  methods:
+    # Without enforce, fallback to user_group
+    pass2.!pass3.!purge_other_groups::
+     "disable_reporting"       usebundle => disable_reporting;
+      "subcall_${group_list}"  usebundle =>  user_group(
```

Comment:
should be done only if user_exists and not group_not_found

Reply:
? i do not understand.
We want to run this only if the 'purge_other_groups' is not defined. The reporting is then delegated to the subcall which also check for group/user existence

Reply:
there's no point calling a submethod if we already know that it will fail